### PR TITLE
platforms/nuttx/src/bootloader/microchip/mpfs/main.c: Improve error h…

### PR DESCRIPTION
**Pull request by jlaitine**
…andling in flashing

Previosly flashing might have succeeded even if the file was not properly
stored in the filesystem

Signed-off-by: Jukka Laitinen <jukkax@ssrc.tii.ae>

### CODE REVIEW

**Comment by jlaitine** [comment]
This moves the end-of flashing check into the flash writing function, to reduce states and complexity, and to enable better error handling.

File: platforms/nuttx/src/bootloader/microchip/mpfs/main.c [hunk 1]
```diff
@@ -491,32 +486,72 @@ flash_func_write_word(uintptr_t address, uint32_t word)
 
 #if defined(CONFIG_MTD_M25P) || defined(CONFIG_MMCSD)
 
-	// Write a single page every time we got a full one
+	static uintptr_t end_address = 0;
```
Comment (pussuw) [inline platforms/nuttx/src/bootloader/microchip/mpfs/main.c:489]: I'm having a hard time understanding why these static variables are needed, also, where do these point if one transfer to disk fails and another attempt is needed ? There is no reset mechanism for these as far as I can see ?

Comment (jlaitine) [reply platforms/nuttx/src/bootloader/microchip/mpfs/main.c:489]: it tries to write to mass media in blocks, but the function is called for every word. So it is collecting the data into DRAM, and keeping track of the progress in the static variables.

end_address is specifically needed in the end of flashing, when less than a full block has been written to the DRAM. Then you need some way to know what is the last word that was written, to be able to write the last incomplete block of data.

**Comment by jlaitine** [comment]
Better than this; we could reserve one block of data from l2lim, and write from there to the disk. Then after all has been flashed, it could just re-trigger the loader task to boot normally from the mass media.

That way the flashing itself doesn't depend on dram, and also you don't need to use the hard coded address for it. Maybe I will continue this PR to make that tomorrow.

File: platforms/nuttx/src/bootloader/microchip/mpfs/main.c (hunk 1 above)
Comment (pussuw) [reply platforms/nuttx/src/bootloader/microchip/mpfs/main.c:489]: Yes exactly, it keeps track of the progress for one process. Good optimization for flash write access but if the process fails / is interrupted, the static variables still point to where the previous process left off. I don't know if this is an issue. Then again, the status of the flash media is also changed and needs erase.

**Comment by pussuw** [comment]
> Better than this; we could reserve one block of data from l2lim, and write from there to the disk. Then after all has been flashed, it could just re-trigger the loader task to boot normally from the mass media.
> 
> That way the flashing itself doesn't depend on dram, and also you don't need to use the hard coded address for it. Maybe I will continue this PR to make that tomorrow.

Maybe I'm just being super nit about this, but for atomicy the static locator variables seem like asking for trouble. At least a reset mechanism for them would be nice, if the transmit must be re-started for whatever reason.

File: platforms/nuttx/src/bootloader/microchip/mpfs/main.c (hunk 1 above)
Comment (jlaitine) [reply platforms/nuttx/src/bootloader/microchip/mpfs/main.c:489]: When a new flashing starts, the static variables are updated again. The write of the new image is always sequential from 0 ->, the end_address is set first time on the second word written, the first_unwritten is set in the end of every write. Both are used only when writing the last word (which is writing the 0 word again with value different than 0xffffffff). So these two don't need reset logic.

**Review by pussuw — APPROVED** [review]

**Review by jpaali — APPROVED** [review]

**Pull request merged**